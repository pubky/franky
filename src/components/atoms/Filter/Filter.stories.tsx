import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { useState } from 'react';

import {
  FilterHeader,
  FilterItem,
  FilterItemIcon,
  FilterItemLabel,
  FilterList,
  FilterRoot,
} from '@/components/atoms/Filter/Filter';
import * as Libs from '@/libs';

const meta: Meta<typeof FilterRoot> = {
  title: 'Atoms/Filter',
  component: FilterRoot,
  subcomponents: {
    FilterHeader,
    FilterList,
    FilterItem,
    FilterItemIcon,
    FilterItemLabel,
  },
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'Foundational filter layout primitives used to assemble consistent filter groups across the product.',
      },
    },
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const RootOnly: Story = {
  name: 'Root Layout',
  render: () => (
    <FilterRoot className="max-w-sm p-4 rounded-xl border bg-background">
      <p className="text-sm text-muted-foreground">Place any filter elements inside this container.</p>
    </FilterRoot>
  ),
  parameters: {
    docs: {
      description: {
        story: 'The base container element that provides consistent spacing and background for filter groups.',
      },
    },
  },
};

export const WithHeaderAndList: Story = {
  name: 'Header + List',
  render: () => (
    <FilterRoot className="max-w-sm p-4 rounded-xl border bg-background">
      <FilterHeader title="Content" />
      <FilterList>
        <FilterItem>
          <FilterItemIcon icon={Libs.Layers} />
          <FilterItemLabel>All</FilterItemLabel>
        </FilterItem>
        <FilterItem>
          <FilterItemIcon icon={Libs.StickyNote} />
          <FilterItemLabel>Posts</FilterItemLabel>
        </FilterItem>
        <FilterItem>
          <FilterItemIcon icon={Libs.Image} />
          <FilterItemLabel>Images</FilterItemLabel>
        </FilterItem>
      </FilterList>
    </FilterRoot>
  ),
  parameters: {
    docs: {
      description: {
        story: 'A composed example combining the header and list primitives to form a complete filter block.',
      },
    },
  },
};

export const InteractiveSelection: Story = {
  name: 'Interactive Selection',
  render: () => {
    const [selected, setSelected] = useState('all');

    const items = [
      { key: 'all', label: 'All', icon: Libs.Layers },
      { key: 'posts', label: 'Posts', icon: Libs.StickyNote },
      { key: 'articles', label: 'Articles', icon: Libs.Newspaper },
    ];

    return (
      <FilterRoot className="max-w-sm p-4 rounded-xl border bg-background">
        <FilterHeader title="Content" />
        <FilterList>
          {items.map(({ key, label, icon: Icon }) => (
            <FilterItem key={key} isSelected={selected === key} onClick={() => setSelected(key)}>
              <FilterItemIcon icon={Icon} />
              <FilterItemLabel>{label}</FilterItemLabel>
            </FilterItem>
          ))}
        </FilterList>
      </FilterRoot>
    );
  },
  parameters: {
    docs: {
      description: {
        story: 'Demonstrates how the primitives work together to highlight the currently selected filter option.',
      },
    },
  },
};
