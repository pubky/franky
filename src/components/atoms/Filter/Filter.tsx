'use client';

import * as React from 'react';
import * as Atoms from '@/atoms';
import * as Libs from '@/libs';

function FilterRoot({ className, children, ...props }: React.HTMLAttributes<HTMLElement>) {
  return (
    <Atoms.Container
      data-slot="filter-root"
      data-testid="filter-root"
      className={Libs.cn('flex flex-col gap-2 bg-background', className)}
      {...props}
    >
      {children}
    </Atoms.Container>
  );
}

function FilterHeader({
  title,
  subtitle,
  className,
  ...props
}: {
  title: string;
  subtitle?: string;
  className?: string;
} & React.HTMLAttributes<HTMLElement>) {
  return (
    <div data-slot="filter-header" data-testid="filter-header" className="flex flex-col gap-2" {...props}>
      <Atoms.Heading level={2} size="lg" className={Libs.cn('text-muted-foreground font-light', className)}>
        {title}
      </Atoms.Heading>
      {subtitle && (
        <p className="text-base font-medium leading-normal text-[var(--base-secondary-foreground,#D4D4DB)]">
          {subtitle}
        </p>
      )}
    </div>
  );
}

function FilterList({ className, children, ...props }: React.HTMLAttributes<HTMLElement>) {
  return (
    <Atoms.Container
      data-slot="filter-list"
      data-testid="filter-list"
      className={Libs.cn('flex flex-col gap-2', className)}
      {...props}
    >
      {children}
    </Atoms.Container>
  );
}

function FilterItem({
  isSelected = false,
  onClick,
  className,
  children,
  ...props
}: {
  isSelected?: boolean;
  onClick?: () => void;
  className?: string;
  children: React.ReactNode;
} & React.HTMLAttributes<HTMLElement>) {
  return (
    <Atoms.Typography
      data-slot="filter-item"
      data-testid="filter-item"
      data-selected={isSelected ? 'true' : 'false'}
      className={Libs.cn(
        'cursor-pointer flex gap-2 items-center text-base font-medium transition-colors',
        isSelected ? 'text-foreground' : 'text-muted-foreground hover:text-foreground/80',
        className,
      )}
      size="sm"
      onClick={onClick}
      {...props}
    >
      {children}
    </Atoms.Typography>
  );
}

function FilterItemIcon({
  icon: Icon,
  className,
  ...props
}: {
  icon: React.ComponentType<{ className?: string }>;
  className?: string;
} & Omit<React.HTMLAttributes<HTMLDivElement>, 'onCopy' | 'onCut' | 'onPaste'>) {
  return <Icon className={Libs.cn('w-5 h-5', className)} {...props} />;
}

function FilterItemLabel({
  children,
  className,
  ...props
}: {
  children: React.ReactNode;
  className?: string;
} & React.HTMLAttributes<HTMLSpanElement>) {
  return (
    <span className={Libs.cn('', className)} {...props}>
      {children}
    </span>
  );
}

export { FilterRoot, FilterHeader, FilterList, FilterItem, FilterItemIcon, FilterItemLabel };
