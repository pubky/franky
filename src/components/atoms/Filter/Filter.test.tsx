import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { FilterRoot, FilterHeader, FilterList, FilterItem, FilterItemIcon, FilterItemLabel } from './Filter';

const MockIcon = ({ className }: { className?: string }) => <svg data-testid="mock-icon" className={className} />;

describe('Filter Components', () => {
  describe('FilterRoot', () => {
    it('renders with default props', () => {
      render(
        <FilterRoot>
          <div>Test content</div>
        </FilterRoot>,
      );

      const root = screen.getByTestId('filter-root');
      expect(root).toBeInTheDocument();
      expect(root).toMatchSnapshot();
    });

    it('applies custom className', () => {
      render(
        <FilterRoot className="custom-class">
          <div>Test content</div>
        </FilterRoot>,
      );

      const root = screen.getByTestId('filter-root');
      expect(root).toMatchSnapshot();
    });
  });

  describe('FilterHeader', () => {
    it('renders with title', () => {
      render(<FilterHeader title="Test Filter" />);

      expect(screen.getByText('Test Filter')).toBeInTheDocument();
      const heading = screen.getByText('Test Filter');
      expect(heading).toMatchSnapshot();
    });

    it('applies custom className', () => {
      render(<FilterHeader title="Test Filter" className="custom-header" />);

      const heading = screen.getByText('Test Filter');
      expect(heading).toMatchSnapshot();
    });
  });

  describe('FilterList', () => {
    it('renders with default props', () => {
      render(
        <FilterList>
          <div>Item 1</div>
          <div>Item 2</div>
        </FilterList>,
      );

      const list = screen.getByTestId('filter-list');
      expect(list).toBeInTheDocument();
      expect(list).toMatchSnapshot();
    });

    it('applies custom className', () => {
      render(
        <FilterList className="custom-list">
          <div>Item 1</div>
        </FilterList>,
      );

      const list = screen.getByTestId('filter-list');
      expect(list).toMatchSnapshot();
    });
  });

  describe('FilterItem', () => {
    it('renders with default props', () => {
      render(
        <FilterItem>
          <span>Test Item</span>
        </FilterItem>,
      );

      const item = screen.getByTestId('filter-item');
      expect(item).toBeInTheDocument();
      expect(item).toMatchSnapshot();
      expect(item).toHaveAttribute('data-selected', 'false');
    });

    it('renders as selected', () => {
      render(
        <FilterItem isSelected={true}>
          <span>Selected Item</span>
        </FilterItem>,
      );

      const item = screen.getByTestId('filter-item');
      expect(item).toMatchSnapshot();
      expect(item).toHaveAttribute('data-selected', 'true');
    });

    it('renders as unselected', () => {
      render(
        <FilterItem isSelected={false}>
          <span>Unselected Item</span>
        </FilterItem>,
      );

      const item = screen.getByTestId('filter-item');
      expect(item).toMatchSnapshot();
      expect(item).toHaveAttribute('data-selected', 'false');
    });

    it('calls onClick when clicked', () => {
      const mockOnClick = vi.fn();
      render(
        <FilterItem onClick={mockOnClick}>
          <span>Clickable Item</span>
        </FilterItem>,
      );

      const item = screen.getByTestId('filter-item');
      fireEvent.click(item);

      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });

    it('applies custom className', () => {
      render(
        <FilterItem className="custom-item">
          <span>Test Item</span>
        </FilterItem>,
      );

      const item = screen.getByTestId('filter-item');
      expect(item).toMatchSnapshot();
    });
  });

  describe('FilterItemIcon', () => {
    it('renders with icon', () => {
      render(<FilterItemIcon icon={MockIcon} />);

      const icon = screen.getByTestId('mock-icon');
      expect(icon).toBeInTheDocument();
      expect(icon).toMatchSnapshot();
    });

    it('applies custom className', () => {
      render(<FilterItemIcon icon={MockIcon} className="custom-icon" />);

      const icon = screen.getByTestId('mock-icon');
      expect(icon).toMatchSnapshot();
    });
  });

  describe('FilterItemLabel', () => {
    it('renders with children', () => {
      render(<FilterItemLabel>Test Label</FilterItemLabel>);

      expect(screen.getByText('Test Label')).toBeInTheDocument();
    });

    it('applies custom className', () => {
      render(<FilterItemLabel className="custom-label">Test Label</FilterItemLabel>);

      const label = screen.getByText('Test Label');
      expect(label).toMatchSnapshot();
    });
  });
});
