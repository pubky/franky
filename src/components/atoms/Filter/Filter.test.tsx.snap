// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filter Components > FilterHeader > applies custom className 1`] = `
<h2
  class="text-2xl text-muted-foreground font-light custom-header"
  data-testid="heading-2"
>
  Test Filter
</h2>
`;

exports[`Filter Components > FilterHeader > renders with title 1`] = `
<h2
  class="text-2xl text-muted-foreground font-light"
  data-testid="heading-2"
>
  Test Filter
</h2>
`;

exports[`Filter Components > FilterItem > applies custom className 1`] = `
<p
  class="cursor-pointer flex gap-2 text-base font-medium text-muted-foreground hover:text-secondary-foreground custom-item"
  data-selected="false"
  data-slot="filter-item"
  data-testid="filter-item"
>
  <span>
    Test Item
  </span>
</p>
`;

exports[`Filter Components > FilterItem > renders as selected 1`] = `
<p
  class="cursor-pointer flex gap-2 text-base font-medium text-foreground"
  data-selected="true"
  data-slot="filter-item"
  data-testid="filter-item"
>
  <span>
    Selected Item
  </span>
</p>
`;

exports[`Filter Components > FilterItem > renders as unselected 1`] = `
<p
  class="cursor-pointer flex gap-2 text-base font-medium text-muted-foreground hover:text-secondary-foreground"
  data-selected="false"
  data-slot="filter-item"
  data-testid="filter-item"
>
  <span>
    Unselected Item
  </span>
</p>
`;

exports[`Filter Components > FilterItem > renders with default props 1`] = `
<p
  class="cursor-pointer flex gap-2 text-base font-medium text-muted-foreground hover:text-secondary-foreground"
  data-selected="false"
  data-slot="filter-item"
  data-testid="filter-item"
>
  <span>
    Test Item
  </span>
</p>
`;

exports[`Filter Components > FilterItemIcon > applies custom className 1`] = `
<div
  class="w-5 h-5 custom-icon"
  data-testid="mock-icon"
>
  Icon
</div>
`;

exports[`Filter Components > FilterItemIcon > renders with icon 1`] = `
<div
  class="w-5 h-5"
  data-testid="mock-icon"
>
  Icon
</div>
`;

exports[`Filter Components > FilterItemLabel > applies custom className 1`] = `
<span
  class="custom-label"
>
  Test Label
</span>
`;

exports[`Filter Components > FilterList > applies custom className 1`] = `
<div
  class="mx-auto w-full flex flex-col gap-2 custom-list"
  data-slot="filter-list"
  data-testid="filter-list"
>
  <div>
    Item 1
  </div>
</div>
`;

exports[`Filter Components > FilterList > renders with default props 1`] = `
<div
  class="mx-auto w-full flex flex-col gap-2"
  data-slot="filter-list"
  data-testid="filter-list"
>
  <div>
    Item 1
  </div>
  <div>
    Item 2
  </div>
</div>
`;

exports[`Filter Components > FilterRoot > applies custom className 1`] = `
<div
  class="mx-auto w-full flex flex-col gap-4 bg-background custom-class"
  data-slot="filter-root"
  data-testid="filter-root"
>
  <div>
    Test content
  </div>
</div>
`;

exports[`Filter Components > FilterRoot > renders with default props 1`] = `
<div
  class="mx-auto w-full flex flex-col gap-4 bg-background"
  data-slot="filter-root"
  data-testid="filter-root"
>
  <div>
    Test content
  </div>
</div>
`;
